Bluetooth: Handle local resolvable private address according to Spec 5.0 text

joerchan opened the pull request:
According to the BT Core spec. The local RPA field in the enhanced
connection complete event should be set if the own-address type was
set to 0x02/0x03 and the controller generated an RPA using a non-zero
IRK, otherwise the controller should return all zeroes.

Our controller does not do this. And this has masked a bug in the host.
If the local RPA is not handled correctly we will not be able to pair as host with controllers
that follow the spec correctly here.
This is the case for Intel controllers.

This pull request addresses both the issue in the Host, and the controller, even though this behavior only complicated the handling for combined build.
It should be fixed though, so that we can test with the correct spec behavior as well as providing spec-compliant controller for controller-only build.

joerchan committed:
Bluetooth: controller: Set local RPA field only when generated

Set the local RPA field of the enhanced connection complete event only
when the address was generated by the controller. If the host has
set an RPA and this one was used the controller should return all
zeroes.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Remove extended advertising from connection complete

Remove handling of extended advertising from connection complete
event. If extended advertising has been enabled and is supported by
the controller then enhanced connection complete must be supported.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Fix compilation issue for extended advertising beacon

Fix compilation issue for extended advertising beacon. Adv pool was
mistakenly put under BT_OBSERVER define.
Both observer and broadcaster use the bt_lookup_id_addr. SMP cannot
be enabled without connections, so for broadcaster and observer there
will not be a lookup.
Fix compilation warning for extended scanner-only build.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Handle zero-set local RPA in enhanced conn complete

According to the BT Core spec. The local RPA field in the enhanced
connection complete event should be set if the own-address type was
set to 0x02/0x03 and the controller generated an RPA using a non-zero
IRK, otherwise the controller should return all zeroes.
In the case where we generate the RPA in the host instead we need to
handle set the on-air RPA address based on the random address set by
the host.
If this is not handled then pairing will fail because the on-air
addresses are used as input to the pairing procedure.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ll_adv.c:
```diff
@@ -443,6 +443,7 @@ u8_t ll_adv_enable(u8_t enable)
 #endif /* CONFIG_BT_CTLR_ADV_EXT */
 	} else {
 		bool priv = false;
+
```
Is this critical change? (just this alone in this file?)

joerchan replied:
Reverted.

joerchan committed:
Bluetooth: controller: Set local RPA field only when generated legacy

Fix local RPA field for the legacy controller. This failed without
compilation errors because of blind pointer cast between two
structs of similar definition.
Set the local RPA field of the enhanced connection complete event only
when the address was generated by the controller. If the host has
set an RPA and this one was used the controller should return all
zeroes.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan committed:
Bluetooth: host: Include RPAs in enhanced conn complete logging

Include the RPA addresses in the enhanced connection complete debug
log. This makes it easier to debug privacy failures.
Use a new debug statement so that the debug print will work correctly
when log_strdup does not copy the string.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

cvinayak reviewed (approved):
Controller changes look good to me.
